### PR TITLE
New specs for a python3 wmagent and wmcore-devtools

### PR DESCRIPTION
**Pull request by amaltaro**
Superseeds https://github.com/cms-sw/cmsdist/pull/5939

For any historical activity, please refer to the initial PR.

### CODE REVIEW

File: wmcorepy3-devtools.spec [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+# Generate dependencies-setup.{sh,csh} so init.{sh,csh} picks full environment.
+rm -rf %i/etc/profile.d
+mkdir -p %i/etc/profile.d
+for tool in $(echo %{requiredtools} | sed -e's|\s+| |;s|^\s+||'); do
+  root=$(echo $tool | tr a-z- A-Z_)_ROOT; eval r=\$$root
+  if [ X"$r" != X ] && [ -r "$r/etc/profile.d/init.sh" ]; then
+    echo "test X\$$root != X || . $r/etc/profile.d/init.sh" >> %i/etc/profile.d/dependencies-setup.sh
+    echo "test X\$?$root = X1 || source $r/etc/profile.d/init.csh" >> %i/etc/profile.d/dependencies-setup.csh
+  fi
+done
```
Comment (smuzaffar) [inline wmcorepy3-devtools.spec]: There is no need to generate these dependencies-setup.* files any more. cmsBuild auto generate them. blocks like above can be removed

File: wmcorepy3-devtools.spec [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+%post
+# The relocation is also needed because of dependencies
+%{relocateConfig}etc/profile.d/dependencies-setup.*sh
```
Comment (smuzaffar) [inline wmcorepy3-devtools.spec]: this is also not needed, cmsBuild takes care of it too

**Comment by muhammadimranfarooqi** [comment]
-1 to b54a10389ab2c3c8f66411c39119344647c9f516 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6021/test_b54a103-20200702-153503.log" target="_blank">here</a>.

File: wmcorepy3-devtools.spec (hunk 2 above)
Comment (amaltaro) [reply wmcorepy3-devtools.spec]: Shahzad, do I need to keep the `%post` statement? Or it can also be removed?

Comment (smuzaffar) [reply wmcorepy3-devtools.spec]: you can remove the `%post` too if there is nothing under it.

**Comment by muhammadimranfarooqi** [comment]
to 0d6c7c5479cdd18597fe0ea26f8cfb5c0d9991dc on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6021/test_0d6c7c5-20200702-212503.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
Classical words, it works in my local machine :)

@muhammadimranfarooqi  Imran, I think we have not updated the correct pkgtools in vocms055. I'd say that the correct place to update pkgtools to the latest commit in that V00-32-XX branch is here:
```
amaltaro@vocms055:/build/dmwmbld/srv $ ll
total 40
drwxr-xr-x.  2 dmwmbld zh 4096 Mar 27  2017 auth
lrwxrwxrwx.  1 dmwmbld zh    7 Mar 30  2017 current -> HG1704e
```

so here this directory needs to be updated:
```
amaltaro@vocms055:/build/dmwmbld/srv/HG1704e/apps.sw.pre/pkgtools $ git log -15 --pretty=oneline --decorate
fad44d44e6bf31ae45828b2f9ddc84bd10f8010f (HEAD, V00-32-XX) Update ChangeLog.md
35c2a7d07cdb64c5cc0aca6ad4c97db98cfc0bd1 Update ChangeLog.md
c09c2ed2b6d63cfc3b8710d3b44c110828aeaf25 Update ChangeLog.md
3614383d317d4c1a005aa3679a3bfcca29cd31a1 Create ChangeLog.md
6771f37cf797988fe0bebd1559d5513ff13c63ef do not download sources for dependency pip packages
```

From the directory structure, I'd even say that this is deployed in some form of COMP RPM or something like that, since I see the HG1704 tag in there.

**Comment by muhammadimranfarooqi** [comment]
@amaltaro 

I have updated pkgtools in the desire location.

**Comment by muhammadimranfarooqi** [comment]
+1 to 0d6c7c5479cdd18597fe0ea26f8cfb5c0d9991dc on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6021/test_0d6c7c5-20200703-095502.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
+1 to 47cbde02608cbdde013f29ee01e4fe8e83bd2b20 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6021/test_47cbde0-20200703-111003.log" target="_blank">here</a>.

**Comment by mapellidario** [comment]
We performed some more investigations regarding the spec files of third-party libraries.

Starting from `wmagentpy3.spec` and `wmcorepy3-devtools.spec`, we considered only `py3-*.spec` entries in the `Requires` field to have `wmagent` and `wmcorepy-devtools` requirements.txt. Then, we installed these requirements in `python 3.6.2` docker environments and used `pip freeze` to get the list of all the packages that have been installed. Every package installed by pip should correspond to a spec file. 

Conclusions: 
* `wmcorepy3-devtools`: everything seems fine
* `wmagentpy3`: `pip freeze` installed two libraries that do not have a spec file yet
  - `importlib-metadata==1.7.0 `
  - `zipp==3.4.0`

[pipdeptree](https://pypi.org/project/pipdeptree/) suggests that `rucio-clients` requires these two packages:

```plaintext
> python -m pip install pipdeptree
> pipdeptree
...
rucio-clients==1.22.2
  - argcomplete [required: >=1.9.0,<1.12.0, installed: 1.11.1]
    - importlib-metadata [required: >=0.23,<2, installed: 1.7.0]
      - zipp [required: >=0.5, installed: 3.4.0]
...
```

Further comment: 
Having used `python 3.6.2`, pip did not install the last version of the packages. For example, last available version of `argcomplete` is `1.12.1`, pip installed `1.11.1` and our spec file has `1.9.4`. Which versions should we use? The ones that @amaltaro chose for the spec files or should we blindly follow pip reccomendations?

More information: 
The `wmagent` and `wmcorepy-devtools` requirements.txt, the dockerfiles, the docker images, the output of `pip freeze` and `pipdeptree` are available on [gitlab.cern.ch](https://gitlab.cern.ch/cms-dmwm-test/wmcore-py3-envs) .

**Comment by amaltaro** [comment]
@mapellidario thank you very much for this wonderful work Dario!

Many of the libraries you reported (through pip freeze command) have a newer version compared to what I have in those spec files. I'm updating all those reported and hopefully nothing will break compared to libraries not present in these lists but necessary to build those services.

In addition to that, I followed your docker procedure and installed the latest `rucio-clients` package, to see what other packages it would bring together, here is the Dockerfile:
```
FROM python:3.6.2

RUN apt-get update && apt-get install -y \
    curl \
    openssl \
    && rm -rf /var/lib/apt/lists/*

RUN python3 -m pip install pip==19.3.1 \
    && python3 -m pip install rucio-clients==1.23.9.post2 
```

and here is the output of pip freeze
```
root@c4f50c2d57e3:/# python3 -m pip freeze
argcomplete==1.11.1
attrs==20.2.0
bcrypt==3.2.0
boto==2.49.0
boto3==1.12.49
botocore==1.15.49
bz2file==0.98
certifi==2020.6.20
cffi==1.14.3
chardet==3.0.4
cryptography==3.2.1
docutils==0.15.2
dogpile.cache==0.6.8
idna==2.10
importlib-metadata==1.7.0
jmespath==0.10.0
jsonschema==3.2.0
paramiko==2.7.2
progressbar2==3.39.3
pycparser==2.20
PyNaCl==1.4.0
pyrsistent==0.17.3
pysftp==0.2.9
python-dateutil==2.8.1
python-magic==0.4.18
python-utils==2.4.0
requests==2.23.0
rucio-clients==1.23.9.post2
s3transfer==0.3.3
six==1.15.0
tabulate==0.8.7
urllib3==1.25.11
zipp==3.4.0
```

rucio-clients will be updated in a second iteration, once I get the current state of this PR validated.
Thanks again!

**Comment by amaltaro** [comment]
@smuzaffar Shahzad, I'm afraid I need your wise advices here again, in case you can spare a few minutes on this.
I pushed a 2nd commit yesterday/today, and cmsBuild gets consistently stuck with those changes.

I have updated pkgtooks to the latest commit on branch V0-32-XX, see:
```
amaltaro@vocms055:/build/amaltaro/pkgtools $ git log -15 --pretty=oneline --decorate
9ac50cd782f34eaa8c688a128fb481d15ac0f5fd (HEAD, V00-32-XX) addlow upload dependnecies via UPLOAD_DEPENDENCIES flag in spec
2c01f518091dbfef8af5ebef511e3982b35b9a32 only use hash tag is tag is actually needed
```

and the problem persists. I have lowered the number of workers to only 1 as well, a command like:
```
amaltaro@vocms055:/build/amaltaro $ pkgtools/cmsBuild -c cmsdist --repository comp    -a slc7_amd64_gcc630 --builders 1 -j 1 --work-dir w    build comp --debug --trace | tee logBuild
```

and still it gets stuck (for more than an hour) building the same `libffi` package. This is the tail of stdout:
```
...
Starting to process package external+py3-pycparser+2.20
Checking if py3-pycparser is cached.
Package external+py3-pycparser+2.20 not found in repository. Queuing for build.
Dependencies for external+py3-pycparser+2.20: ['install-external+gcc+6.3.0', 'install-external+py3-pip+19.3.1', 'install-external+python3+3.6.2', 'install-external+py3-setuptools+39.2.0', 'fetch-external+py3-pycparser+2.20']
Starting to process package external+libffi+3.2.1
Checking if libffi is cached.
libffi is available from remote repository "comp". Not building. Downloading and installing...
About to install external+libffi+3.2.1 using cmspkg...
Done installing via cmspkg.
Package external+libffi+3.2.1 found in repository
```

As you can see, I have both debug and trace options enabled, but still no clue of where exactly it gets stuck.
Would you have any suggestion here? Should I add some extra debugging to `cmsBuild`? Or you think the problem might be somewhere else?

**Comment by smuzaffar** [comment]
@amaltaro , normally for new packages,  I start with building one package at a time. cmsBuild then complains about missing dependencies and I keep on fixing :-) this way Only needed dependencies are added in the correct specs . Can you try this approch?

**Comment by amaltaro** [comment]
Sure, right on! Thanks Shahzad!

**Comment by smuzaffar** [comment]
@amaltaro , lets say you want to add `rucio-client` then just add `rucio-client` in requirements.txt and build. cmsBuild will complain about missing dependencies. Then just added those missing dependencies in requirements.txt only and build them one by one to find out their individual dependencies

**Comment by amaltaro** [comment]
No luck! Even if I try to build only `python3`, it keeps getting stuck while running one of the hundreds of profiling tests like:
```
Running code to generate profile data (this can take a while):
make run_profile_task
make[1]: Entering directory `/build/amaltaro/w/BUILD/slc7_amd64_gcc630/external/python3/3.6.2/Python-3.6.2'
: # FIXME: can't run for a cross build
LD_LIBRARY_PATH=/build/amaltaro/w/BUILD/slc7_amd64_gcc630/external/python3/3.6.2/Python-3.6.2:/build/amaltaro/w/slc7_amd64_gcc630/external/zlib/1.2.8/lib:/build/amaltaro/w/slc7_amd64_gcc630/external/sqlite/3.8.10.2/lib:/build/amaltaro/w/slc7_amd64_gcc630/external/readline/6.2/lib:/build/amaltaro/w/slc7_amd64_gcc630/external/openssl/1.0.1r-comp/lib:/build/amaltaro/w/slc7_amd64_gcc630/external/ncurses/6.0/lib:/build/amaltaro/w/slc7_amd64_gcc630/external/gdbm/1.12/lib:/build/amaltaro/w/slc7_amd64_gcc630/external/expat/2.0.1/lib:/build/amaltaro/w/slc7_amd64_gcc630/external/db4/4.8.30/lib:/build/amaltaro/w/slc7_amd64_gcc630/external/bz2lib/1.0.6/lib:/build/amaltaro/w/slc7_amd64_gcc630/external/gcc/6.3.0/lib64:/build/amaltaro/w/slc7_amd64_gcc630/external/gcc/6.3.0/lib:/build/amaltaro/w/slc7_amd64_gcc630/external/rpm/4.13.0.1/lib:/build/amaltaro/w/slc7_amd64_gcc630/external/bootstrap-bundle/1.7/lib ./python -m test.regrtest --pgo || true
Run tests sequentially
0:00:00 load avg: 2.96 [  1/405] test_grammar
...
0:13:09 load avg: 0.81 [198/405] test_logging
```
and remains in the test above forever. I'm not sure whether this test/profiling is mandatory or not, neither I was able to pin point this option from the `configure` file.

BTW, this problem happens with either pkgtools commits (current vs HEAD).

My next step will be to try to upgrade python to the same version/spec as in the CMSSW branch (3.8.2), without all the spec changes that I have here...

**Comment by muhammadimranfarooqi** [comment]
-1 to b50407f60290b080fa33418949dc6a28f0d8638a on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6021/test_b50407f-20201104-231503.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to 4c8824a9daf871e1a8a6a75d96bf9cc2e0204638 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6021/test_4c8824a-20201106-075005.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to b34236f66295a23668ce99c95d8cf8414319e113 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6021/test_b34236f-20201106-115504.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
It seems to fail to fetch the pypi packages through https... this is the error from the log above:
```
Error while downloading sources from pip://pep8/1.7.0?pip_options=--no-deps%20--no-binary%3D:all:&pip=pip3&package_dependency=py3-pip&output=/source.tar.gz.

cd /build/dmwmbld/srv/state/dmwmbld/builds/6021/w/tmp/tmpQPVtAZ; . /build/dmwmbld/srv/state/dmwmbld/builds/6021/w/slc7_amd64_gcc630/external/py3-pip/19.3.1/etc/profile.d/init.sh; pip3 download --no-cache-dir  --no-deps --no-binary=:all: --disable-pip-version-check  -d . pep8==1.7.0; /bin/ls | grep -v files.list > files.list 
256 
WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/pep8/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/pep8/
WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/pep8/
WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/pep8/
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/pep8/
Could not fetch URL https://pypi.org/simple/pep8/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.org', port=443): Max retries exceeded with url: /simple/pep8/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
```

Not sure whether this is something new from the latest python update (to 3.8.2) or something else...

**Comment by amaltaro** [comment]
This PR has been superseeded by a few other PRs that have been already merged.

**Pull request closed without merging**